src/validation.cpp

```diff
@@ -2566,6 +2567,16 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
     int nStopAtHeight = gArgs.GetArg("-stopatheight", DEFAULT_STOPATHEIGHT);
     do {
         boost::this_thread::interruption_point();
+
+        if (GetMainSignals().CallbacksPending() > 10) {
+            // Block until the validation queue drains
+            std::promise<void> promise;
+            CallFunctionInValidationInterfaceQueue([&promise] {
+                promise.set_value();
+            });
+            promise.get_future().wait();
```

Comment:
In commit "Block ActivateBestChain to empty validationinterface queue"

Could add a CValidationInterface::WaitForPendingCallbacks() method to wrap this up since this is also needed in other places.

Reply:
Done, in a new commit.